Fix solver logic to QF_LIA

senier opened the pull request:
# Description

~~This just fixes the solver logic to `QF_LIA` without changing types bit vectors as planned originally. For checking all example specifications we observe a 50-fold speedup. I have an initial conversion to bitvectors, but there were still thirty failing tests. I don't think this big change is justified given the speedup we get by just fixing the logic.~~

EDIT: I forgot to clean the cache. We also need `QF_NIA`, as `QF_LIA` does not support `*`, `/`, `mod` etc. The actual speedup is between 3% and 11%. Consequently, the BV changes  in #612 and #625 may still be useful.

@treiher Please try with your DHCP example.

treiher commented:
I get a long list of errors for DHCP after this change:

```
specs/dhcp.rflx:163:18: model: error: unreachable field "Len" in "DHCP::Option"
specs/dhcp.rflx:163:18: model: info: path 0 (Code -> Len):
specs/dhcp.rflx:180:18: model: error: unreachable field "Cookie_Servers" in "DHCP::Option"
specs/dhcp.rflx:180:18: model: info: path 0 (Code -> Len -> Cookie_Servers):
specs/dhcp.rflx:176:18: model: error: unreachable field "Domain_Name_Servers" in "DHCP::Option"
specs/dhcp.rflx:176:18: model: info: path 0 (Code -> Len -> Domain_Name_Servers):
specs/dhcp.rflx:184:18: model: error: unreachable field "Impress_Servers" in "DHCP::Option"
specs/dhcp.rflx:184:18: model: info: path 0 (Code -> Len -> Impress_Servers):
specs/dhcp.rflx:182:18: model: error: unreachable field "LPR_Servers" in "DHCP::Option"
specs/dhcp.rflx:182:18: model: info: path 0 (Code -> Len -> LPR_Servers):
specs/dhcp.rflx:178:18: model: error: unreachable field "Log_Servers" in "DHCP::Option"
specs/dhcp.rflx:178:18: model: info: path 0 (Code -> Len -> Log_Servers):
specs/dhcp.rflx:174:18: model: error: unreachable field "Name_Servers" in "DHCP::Option"
specs/dhcp.rflx:174:18: model: info: path 0 (Code -> Len -> Name_Servers):
specs/dhcp.rflx:256:18: model: error: unreachable field "NetBIOS_Over_TCP_IP_Datagram_Distribution_Servers" in "DHCP::Option"
specs/dhcp.rflx:256:18: model: info: path 0 (Code -> Len -> NetBIOS_Over_TCP_IP_Datagram_Distribution_Servers):
specs/dhcp.rflx:254:18: model: error: unreachable field "NetBIOS_Over_TCP_IP_Name_Servers" in "DHCP::Option"
specs/dhcp.rflx:254:18: model: info: path 0 (Code -> Len -> NetBIOS_Over_TCP_IP_Name_Servers):
specs/dhcp.rflx:248:18: model: error: unreachable field "Network_Information_Servers" in "DHCP::Option"
specs/dhcp.rflx:248:18: model: info: path 0 (Code -> Len -> Network_Information_Servers):
specs/dhcp.rflx:250:18: model: error: unreachable field "Network_Time_Protocol_Servers" in "DHCP::Option"
specs/dhcp.rflx:250:18: model: info: path 0 (Code -> Len -> Network_Time_Protocol_Servers):
specs/dhcp.rflx:214:18: model: error: unreachable field "Path_MTU_Plateau_Table" in "DHCP::Option"
specs/dhcp.rflx:214:18: model: info: path 0 (Code -> Len -> Path_MTU_Plateau_Table):
specs/dhcp.rflx:206:18: model: error: unreachable field "Policy_Filter" in "DHCP::Option"
specs/dhcp.rflx:206:18: model: info: path 0 (Code -> Len -> Policy_Filter):
specs/dhcp.rflx:186:18: model: error: unreachable field "Resource_Location_Servers" in "DHCP::Option"
specs/dhcp.rflx:186:18: model: info: path 0 (Code -> Len -> Resource_Location_Servers):
specs/dhcp.rflx:170:18: model: error: unreachable field "Routers" in "DHCP::Option"
specs/dhcp.rflx:170:18: model: info: path 0 (Code -> Len -> Routers):
specs/dhcp.rflx:230:18: model: error: unreachable field "Static_Routes" in "DHCP::Option"
specs/dhcp.rflx:230:18: model: info: path 0 (Code -> Len -> Static_Routes):
specs/dhcp.rflx:172:18: model: error: unreachable field "Time_Servers" in "DHCP::Option"
specs/dhcp.rflx:172:18: model: info: path 0 (Code -> Len -> Time_Servers):
specs/dhcp.rflx:457:29: parser: error: undefined element type "DHCP::Option"
specs/dhcp.rflx:480:10: model: error: missing type for field "Options" in "DHCP::Message"
```

The specification can be found [here](https://github.com/Componolit/RecordFlux-specifications/blob/issue_2/dhcp.rflx).

senier commented:
> I get a long list of errors for DHCP after this change:

This was due to using the wrong logic. I changed this in 73431da9. I also added slightly more useful error messages in those cases in e2ef93ce. See the updated PR description - the performance improvement is not at all as impressive as I thought.

treiher commented:
> Consequently, the BV changes in #612 and #625 may still be useful.

The commit message of 73431da9c34a60c23ec9dbae55c798170fa4070c still closes both issues. Is that intended?

treiher commented on rflx/expression.py:
```diff
@@ -40,26 +40,33 @@ class ProofResult(Enum):
     def result(self) -> ProofResult:
         return self.__result
 
     @property
     def error(self) -> List[Tuple[str, Optional[Location]]]:
+        if self.__result == ProofResult.UNKNOWN:
+            return [(self.__unknown_reason, None)]
```
Would it make sense to return the location of `self.__expr` to have at least any location in this case?

senier replied:
No, I tried it with the DHCP example which made me aware of the issue and it gave me no location either. We use `expr.proof(facts)` most of the time where `expr` is constructed (or just `TRUE`) and thus has no location.

treiher replied:
I see.

treiher commented on rflx/expression.py:
```diff
@@ -40,26 +40,33 @@ class ProofResult(Enum):
 
         self.__result = ProofResult(solver.check())
+        if self.__result == ProofResult.UNKNOWN:
+            self.__unknown_reason = solver.reason_unknown()
```
Couldn't this be done in the `error` property? The unsat core is also analyzed there. Why should `UNKOWN` handled differently than `UNSAT`?

senier replied:
For `UNSAT` we rerun the proof completely with unsat core generation enabled (as the original solver object just doesn't have this information). Do we really want to rerun the proof for `UNKNOWN` just to get the same information we already had in the previous proof attempt?

treiher replied:
No, this doesn't make sense. I didn't realize that rerunning the proof would be necessary.

treiher commented on rflx/expression.py:
```diff
@@ -40,26 +40,33 @@ class ProofResult(Enum):
     def result(self) -> ProofResult:
         return self.__result
 
     @property
     def error(self) -> List[Tuple[str, Optional[Location]]]:
+        if self.__result == ProofResult.UNKNOWN:
+            return [(self.__unknown_reason, None)]
         assert self.__result == ProofResult.UNSAT
```
What about returning an empty list in case of `SAT` instead of having the assert?

senier replied:
You're not supposed to call error in case of `SAT`. I dislike the idea of hiding these kind of usage errors by returning an empty list, as they often hint another problem (e.g. you forgot to properly check the proof result).

senier commented:
> > Consequently, the BV changes in #612 and #625 may still be useful.
> 
> The commit message of [73431da](https://github.com/Componolit/RecordFlux/commit/73431da9c34a60c23ec9dbae55c798170fa4070c) still closes both issues. Is that intended?

No, changed in e84a8ee6.

treiher commented on rflx/expression.py:
```diff
@@ -40,26 +40,33 @@ class ProofResult(Enum):
         self.__expr = expr
         self.__facts = facts or []
         self.__result = ProofResult.UNSAT
+        self.__unknown_reason = "unknown"
```
Is there any case where the default value of `__unknown_reason` is used? If not, I think it would be better to use an empty string as default value to prevent the impression that it is relevant in some case.

senier replied:
No, it's only used when set to an error message iff the result is `UNKNOWN`. I made it an optional `str` type and set it to `None` in ea507e4c127f31ec08e31eb5f56f823638c9d191.

senier commented:
@treiher I had to refactor the PR a bit to achieve test coverage. The logic became a parameter to the constructor of the `Proof` class so we can set it to an unsuitable logic in the tests and trigger the `UNKNOWN` issue.

senier commented:
@treiher Rebased to current `main`.
